src/mito2/src/engine/truncate_test.rs

```diff
@@ -0,0 +1,441 @@
+#[tokio::test]
+async fn test_engine_truncate_reopen() {
+    let mut env = TestEnv::with_prefix("truncate-reopen");
+    let engine = env.create_engine(MitoConfig::default()).await;
+
+    // Create the region.
+    let region_id = RegionId::new(1, 1);
+    let request = CreateRequestBuilder::new().build();
+    let region_dir = request.region_dir.clone();
+
+    let column_schemas = rows_schema(&request);
+    engine
+        .handle_request(region_id, RegionRequest::Create(request))
+        .await
+        .unwrap();
+
+    // Put data to the region.
+    let rows = Rows {
+        schema: column_schemas.clone(),
+        rows: build_rows(0, 3),
+    };
+    put_rows(&engine, region_id, rows).await;
+
+    let region = engine.get_region(region_id).unwrap();
+    let last_entry_id = region.version_control.current().last_entry_id;
+
+    // Truncate the region
+    engine
+        .handle_request(region_id, RegionRequest::Truncate(RegionTruncateRequest {}))
+        .await
+        .unwrap();
+
+    // Reopen the region again.
+    let engine = env.reopen_engine(engine, MitoConfig::default()).await;
+    engine
+        .handle_request(
+            region_id,
+            RegionRequest::Open(RegionOpenRequest {
+                engine: String::new(),
+                region_dir,
+                options: HashMap::default(),
+            }),
+        )
+        .await
+        .unwrap();
+
+    let region = engine.get_region(region_id).unwrap();
+    assert_eq!(last_entry_id, region.version().flushed_entry_id);
+
+    // Scan the region.
+    let request = ScanRequest::default();
+    let stream = engine.handle_query(region_id, request).await.unwrap();
+    let batches = RecordBatches::try_collect(stream).await.unwrap();
+    let expected = "++\n++";
+    assert_eq!(expected, batches.pretty_print().unwrap());
+}
+
+#[tokio::test]
+async fn test_engine_truncate_during_flush() {
+    let mut env = TestEnv::with_prefix("truncate-during-flush");
+    let engine = env.create_engine(MitoConfig::default()).await;
+
+    // Create the region.
+    let region_id = RegionId::new(1, 1);
+    let request = CreateRequestBuilder::new().build();
+    let region_dir = request.region_dir.clone();
+
+    let column_schemas = rows_schema(&request);
+    engine
+        .handle_request(region_id, RegionRequest::Create(request))
+        .await
+        .unwrap();
+
+    // Put data to the region.
+    let rows = Rows {
+        schema: column_schemas.clone(),
+        rows: build_rows(0, 3),
+    };
+    put_rows(&engine, region_id, rows).await;
+
+    let region = engine.get_region(region_id).unwrap();
+
+    // Create a parquet file.
+    // Simulate that the `do_flush()` function is currently being executed.
+    let file_id = FileId::random();
+    let file_name = format!("{}.parquet", file_id);
+    let file_meta = FileMeta {
+        region_id,
+        file_id,
+        time_range: FileTimeRange::default(),
+        level: 0,
+        file_size: 0,
+    };
```

Comment:
suggestion: You could add an event to the `EventListener` and `WorkerListener`
https://github.com/GreptimeTeam/greptimedb/blob/50cea29b29a45073a11b1cb3702a5eaa21b465cb/src/mito2/src/worker.rs#L556-L559

The event is something like `on_flush_begin()` and we call it before flushing memtables
https://github.com/GreptimeTeam/greptimedb/blob/50cea29b29a45073a11b1cb3702a5eaa21b465cb/src/mito2/src/flush.rs#L221-L223

Then you could implement a listener that blocks the flush job in the background.
- spawns a task to trigger the flush
  - it will block on `on_flush_begin()`
  - returns flush result
- truncates the region
- notifies the listener
- joins the spawned task to get the result, checks the result
- checks sequences and entry ids


We don't need to expose and call `handle_worker_request()`.